File: test/test_recall_image.py
import sys

import torch
import cv2 as cv
import numpy as np

from yoro.visual import rbox_draw
# from yoro.api import non_maximum_suppression as nms

if __name__ == '__main__':

    if len(sys.argv) < 3:
        print('Usage: %s <model_path> <image_path>' % sys.argv[0])
        exit()

    modelPath = sys.argv[1]
    imgPath = sys.argv[2]

    # Load model and image
    model = torch.jit.load(modelPath)
    netWidth = model.suffix.width
    netHeight = model.suffix.height
    model.eval()

    img = cv.imread(imgPath, cv.IMREAD_COLOR)

    # Padding image to square
    height, width, channel = img.shape
    tarSize = max(height, width)

    startX = int((tarSize - width) / 2)
    startY = int((tarSize - height) / 2)

    mat = np.zeros((tarSize, tarSize, channel), dtype=img.dtype)
    mat[startY:startY+height, startX:startX+width, :] = img
    mat = cv.resize(mat, (netWidth, netHeight))
    mat = cv.cvtColor(mat, cv.COLOR_BGR2RGB)
    mat = mat.transpose(2, 0, 1)

    # Predict
    inputs = (torch.FloatTensor(mat) / 255).unsqueeze(0)
    outputs = model(inputs)

    # Convert results
    results = []
    confTh = 0.7

    shift = torch.tensor(
        [[[startX, startY]]], dtype=inputs.dtype, device=inputs.device)
    for i, (conf, label, boxes, degree) in enumerate(outputs):

        # Denormalize
        boxes.mul_(tarSize / netWidth)
        boxes[..., 0:2].sub_(shift)

        # Convert result
        results += [
            {
                'label': label[n, a, h, w].item(),
                'x': boxes[n, a, h, w, 0].item(),
                'y': boxes[n, a, h, w, 1].item(),
                'w': boxes[n, a, h, w, 2].item(),
                'h': boxes[n, a, h, w, 3].item(),
                'degree': degree[n, a, h, w].item()
            }
            for n in range(conf.size(0))
            for a in range(conf.size(1))
            for h in range(conf.size(2))
            for w in range(conf.size(3))
            if conf[n, a, h, w] >= confTh
        ]

    for inst in results:
        print(inst)

    # Draw result
    result = rbox_draw([img], [results])
    cv.imshow('result', result[0])
    cv.waitKey(0)

    """
    # Apply non-maximum suppression
    confTh = 0.9
    nmsTh = 0.7
    nmsOut = nms(outputs, confTh, nmsTh)

    # Draw result
    print(nmsOut)
    result = rbox_draw([img], nmsOut)
    cv.imshow('result', result[0])
    cv.waitKey(0)
    """

File: test/test_recall_image.cpp
#include <cstdio>
#include <exception>
#include <iostream>
#include <memory>
#include <vector>

#include <torch/script.h>
#include <opencv2/opencv.hpp>

#include <calc_ops.hpp>

using namespace std;
using namespace cv;

using torch::from_blob;
using torch::ScalarType;
using torch::Tensor;
using torch::indexing::Slice;
using torch::jit::load;
using torch::jit::Object;
using torch::jit::script::Module;

int main(int argc, char* argv[])
try
{
    // Check arguments
    if (argc < 3)
    {
        printf("Usage: %s <model_path> <test_image>\n", argv[0]);
        return -1;
    }

    // Import model and settings
    Module model = load(argv[1]);
    Object yoroLayer = model.attr("yoroLayer").toObject();

    int netWidth = yoroLayer.attr("width").toInt();
    int netHeight = yoroLayer.attr("height").toInt();

    model.eval();

    // Read image
    Mat src = imread(argv[2], IMREAD_COLOR);
    cvtColor(src, src, COLOR_BGR2RGB);

    // Pad to square
    int tarSize = max(src.rows, src.cols);
    Mat mat = Mat(tarSize, tarSize, CV_8UC3, Scalar(0, 0, 0));
    Rect roi = Rect(
        (tarSize - src.cols) / 2, (tarSize - src.rows) / 2, src.cols, src.rows);
    src.copyTo(mat(roi));

    float startX = (tarSize - src.cols) / 2;
    float startY = (tarSize - src.rows) / 2;
    float scale = float(tarSize) / float(netWidth);

    // Resizing
    resize(mat, mat, Size(netWidth, netHeight));

    // Convert image to tensor
    Tensor inputs =
        from_blob(
            mat.ptr<char>(), {1, netHeight, netWidth, 3}, ScalarType::Byte)
            .to(torch::kFloat)
            .permute({0, 3, 1, 2})
            .contiguous() /
        255.0;

    // Forward
    auto outputs = model.forward({inputs}).toTuple();
    auto listRef = outputs->elements();
    Tensor predConf = listRef[0].toTensor();
    Tensor predClass = listRef[1].toTensor();
    Tensor predClassConf = listRef[2].toTensor();
    Tensor predBox = listRef[3].toTensor();
    Tensor predDeg = listRef[4].toTensor();

    // Denormalize
    predBox *= scale;
    predBox.index({"...", Slice(0, 2)}) -= torch::tensor({{{startX, startY}}});

    // Concatenate tensor
    Tensor pred = torch::cat(
                      {predConf.unsqueeze(-1).to(torch::kFloat),
                       predClass.unsqueeze(-1).to(torch::kFloat),
                       predClassConf.unsqueeze(-1),
                       predDeg.unsqueeze(-1),
                       predBox},
                      2)
                      .to(torch::kCPU);

    // Processing non-maximum suppression
    vector<vector<yoro_api::RBox>> nmsOut =
        yoro_api::non_maximum_suppression(pred, 0.9, 0.7);
    for (size_t n = 0; n < nmsOut.size(); n++)
    {
        cout << "Batch " << n << ":" << endl;
        for (size_t i = 0; i < nmsOut[n].size(); i++)
        {
            cout << nmsOut[n][i].to_string() << endl;
        }
        cout << endl;
    }

    return 0;
}
catch (exception& ex)
{
    cout << "Error occurred:" << endl;
    cout << ex.what() << endl;
    return -1;
}
